proctor-common/src/main/java/com/indeed/proctor/common/TestRangeSelector.java

```diff
@@ -117,10 +133,12 @@ private int findMatchingRuleInternal(
                     return -1;
                 }
             }
-
             for (int i = 0; i < rules.length; i++) {
                 rule = rules[i];
-                if (evaluator.apply(rule)) {
+                if (identifier != null
```

Comment:
is the extra check for missingExperimentUnit necessary?
The calling code is responsible to set missingExperimentUnit in the context. Is that correct?

Reply:
Yes it is still required. For more information view this test: https://github.com/indeedeng/proctor/pull/130/files/29ac466519631c75b05b563ed70e6f2c9344459f#diff-731be0a9c608b5ec73b9e68c6bef7186fd997e7adfbd6b33caec1d3ece097e48R839 

The context is set but if no rule is matched with missingExperimentalUnit, you should still not evaluate the rules without missingExperimentalUnit.

Reply:
Sorry i am still trying to understand. 
if missingExperimentalUnit is set and other condition is met, will evaluator.apply(rule) just return true?
if missingExperimentalUnit is not set, the rule will evaluate as usual still.

Is that correct?

Reply:
Since normally this section of code is not hit for invalid Identifiers we still need to block it for rules without the missingExperimentalUnit. For example:

- missingExperimentalUnit && country == 'us'
- country == 'us' && lang == 'en'

If invalid identifier is entered we still want to evaluate the rules with missingExperimentalUnit only. This if statement allows us to only match allocations when either have a valid identifier or contain missingExperimentalUnit.  

The missingExperimentalUnit will be evaluated in both cases but when a valid identifier is present it is set to false in the context.

Reply:
now i understand. Thanks Zack